utils/funcexp/func_quote.cpp

```diff
@@ -0,0 +1,105 @@
+/* Copyright (C) 2019 MariaDB Corporation.
+
+   This program is free software; you can redistribute it and/or
+   modify it under the terms of the GNU General Public License
+   as published by the Free Software Foundation; version 2 of
+   the License.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program; if not, write to the Free Software
+   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+   MA 02110-1301, USA. */
+
+#include <string>
+using namespace std;
+
+#include "functor_str.h"
+using namespace execplan;
+
+#include "rowgroup.h"
+using namespace rowgroup;
+
+#define get_esc_bit(mask, num) (1 & (*((mask) + ((num) >> 3))) >> ((num) & 7))
+
+namespace funcexp
+{
+
+CalpontSystemCatalog::ColType Func_quote::operationType(FunctionParm& fp, CalpontSystemCatalog::ColType& resultType)
+{
+    return resultType;
+}
+
+std::string Func_quote::getStrVal(rowgroup::Row& row,
+                                  FunctionParm& fp,
+                                  bool& isNull,
+                                  execplan::CalpontSystemCatalog::ColType& op_ct)
+{
+    string str;
+
+    stringValue(fp[0], row, isNull, str);
+
+    if (isNull)
+    {
+        isNull = false;
+        return "NULL";
+    }
+
+    if (str.empty())
+        return "NULL";
+
+    unsigned char bit_mask[32] = {0x01, 0x00, 0x00, 0x04, 0x08, 0x00, 0x00, 0x00,
+                                  0x00, 0x00, 0x00, 0x10, 0x00, 0x00, 0x00, 0x00,
+                                  0x00, 0x00, 0x00, 0x10, 0x00, 0x00, 0x00, 0x00,
+                                  0x00, 0x00, 0x00, 0x10, 0x00, 0x00, 0x00, 0x00};
+
+    uint64_t arg_length = str.size(), new_length = arg_length + 2;
+
+    for (char ch : str)
+    {
+      new_length += get_esc_bit(bit_mask, (unsigned char) ch);
+    }
+
+    char *result = new char[new_length + 1];
+    char *to = result;
+
+    *to++ = '\'';
+    for (uint64_t i = 0; i < str.size(); i++, to++)
```

Comment:
You have two loops over the argument chars. Could you reduce a number of loops over the argument to one?

Reply:
Changes made as discussed. Please review.